### PR TITLE
Jk initialization

### PR DESCRIPTION
**Pull request by kuba-krj**
# Description
Tidy up initialization in our repo and add option of truncated normal

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary Link: [Neptune](https://app.neptune.ai/o/pmtest/org/llm-random/runs/table?viewId=9a47abb5-019a-4bbc-8109-a30a95af13c3&detailsTab=metadata&shortId=LLMRANDOM-2067&type=run&compare=EwBgzAnANAjFxlkuMg&lbViewUnpacked=true&sortBy=%5B%22sys%2Fcreation_time%22%5D&sortFieldType=%5B%22datetime%22%5D&sortFieldAggregationMode=%5B%22auto%22%5D&sortDirection=%5B%22descending%22%5D&suggestionsEnabled=false&query=%28sys%2Fid%3Astring+%3D+%22LLMRANDOM-2067%22%29)
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review

### CODE REVIEW

File: research/timing/bert_time.py [hunk 1]
```diff
@@ -282,7 +273,7 @@ def main_tests(version, disable_inner=False, expertsets=4, expertsize=64, nexper
             ),
         )
     elif version == "dense":
-        sparse_linear_projection = lambda: misc.DenseEinMix(dm, dm)
+        sparse_linear_projection = lambda: modules.DenseEinMix(dm, dm)
```
Comment (crewtool) [inline research/timing/bert_time.py:276]: Why rename?

File: research/nonlinearities/core/research_bert.py [hunk 2]
```diff
@@ -2,10 +2,11 @@
 
 import numpy as np
 import torch
+from lizrd.core.misc import EinMix
 
 import lizrd.core.nn as nn
 from lizrd.core.llm import LowRank
-from lizrd.core.misc import EinMix, Linear
+from lizrd.core.misc import Linear
```
Comment (crewtool) [inline research/nonlinearities/core/research_bert.py:9]: imo tohether looks better

File: research/conditional/tests/test_token_choice.py [hunk 3]
```diff
@@ -1,9 +1,10 @@
 import torch
 from torch.nn import Sequential, ReLU
+from lizrd.core.misc import Linear
 
 from research.conditional.moe_layers.token_choice import TokenChoiceFF
 from lizrd.support.test_utils import GeneralTestCase
-from lizrd.core.misc import Linear, propagate_forward_pass_cache
+from lizrd.core.misc import propagate_forward_pass_cache
```
Comment (crewtool) [inline research/conditional/tests/test_token_choice.py:7]: better togehther

File: research/conditional/moe_layers/ffs.py [hunk 4]
```diff
@@ -519,11 +528,13 @@ def __init__(self, dinput, doutput, modules):
         assert doutput % modules == 0
         dmodule = doutput // modules
 
-        self.gating = nn.Parameter(misc.get_init_weight((modules, dinput), fan_in=1))
+        self.gating = nn.Parameter(
+            lizrd.core.init.get_init_weight((modules, dinput), fan_in=1)
```
Comment (crewtool) [inline research/conditional/moe_layers/ffs.py:532]: maybe add import get_init_weight from  lizrd.core.init

File: research/conditional/moe_layers/cont_moe_designs/separate_weighted_parameter_matrices.py [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 import torch
 from lizrd.core import misc, nn
+import lizrd.core.init
```
Comment (crewtool) [inline research/conditional/moe_layers/cont_moe_designs/separate_weighted_parameter_matrices.py:5]: mayhbe import get_init_weight

File: research/conditional/moe_layers/cont_moe_designs/separate_merge_emit_weights_common_base.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 from lizrd.core import misc, nn
+import lizrd.core.init
```
Comment (crewtool) [inline research/conditional/moe_layers/cont_moe_designs/separate_merge_emit_weights_common_base.py:2]: maybe import get_init_weight

File: research/conditional/moe_layers/cont_moe_designs/separate_merge_emit_weights.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
 from lizrd.core import nn, misc
+import lizrd.core.init
```
Comment (crewtool) [inline research/conditional/moe_layers/cont_moe_designs/separate_merge_emit_weights.py:2]: maybe import get_init_weight

File: lizrd/core/llm.py [hunk 8]
```diff
@@ -374,15 +357,43 @@ def get_current_device(self, block_num):
...
+):
+    embedding = nn.Embedding(vocab_size, embedding_dim)
+    default_weight = embedding.weight.data
+    embedding.weight.data = get_init_weight(
+        shape=default_weight.shape,
+        fan_in=1,
+        init_type=init_type,
+        scale=init_scale,
+        dtype=default_weight.dtype,
+    )
```
Comment (crewtool) [inline lizrd/core/llm.py:374]: Maybe pass weight as a parameter

File: lizrd/core/init.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    elif init_type == "truncated_normal":
+        return init_truncated_normal(
+            shape=shape, fan_in=fan_in, scale=scale, dtype=dtype
+        )
+    else:
+        raise ValueError(f"Unknown init_type: {init_type}")
+
+
+def init_uniform(shape, fan_in, gain=1.0, dtype=torch.float32):
+    range_ = gain * (3 / fan_in) ** 0.5
```
Comment (crewtool) [inline lizrd/core/init.py:17]: Is this 3 important? Why 3 anyway?

File: lizrd/core/init.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    return torch.zeros(shape, dtype=dtype).uniform_(-range_, range_)
+
+
+def init_truncated_normal(shape, fan_in, scale=0.1, dtype=torch.float32):
```
Comment (crewtool) [inline lizrd/core/init.py:21]: Deafult value for scale should not be set

File: lizrd/core/init.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    t = torch.zeros(shape, dtype=dtype)
+    return trunc_normal_(t, mean=0.0, std=std, a=low, b=high)
+
+
+def get_init_bias(shape, fan_in=None, fan_out=None, dtype=torch.float32):
+    if fan_in is not None:
+        raise ValueError("fan_in unsupported")
+    if fan_out is not None:
+        raise ValueError("fan_out unsupported")
+    return torch.zeros(shape, dtype=dtype)
```
Comment (crewtool) [inline lizrd/core/init.py:34]: It actually returns tensor of zeros xD

**Review by crewtool — APPROVED** [review]
LGTM!
Added small comments though

File: lizrd/core/init.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import torch
+from torch.nn.init import trunc_normal_
+
+
+def get_init_weight(shape, fan_in, init_type="uniform", scale=1.0, dtype=torch.float32):
+    if init_type == "uniform":
```
Comment (Simontwice) [inline lizrd/core/init.py:6]: can we change to the actual name of the initialisation? Like Kaiming, Glorot or whatever this particular one is

File: lizrd/core/init.py (hunk 10 above)
Comment (Simontwice) [reply lizrd/core/init.py:21]: agreed

File: lizrd/core/init.py (hunk 11 above)
Comment (Simontwice) [reply lizrd/core/init.py:34]: @crewtool as it should

File: lizrd/core/init.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import torch
+from torch.nn.init import trunc_normal_
+
+
+def get_init_weight(shape, fan_in, init_type="uniform", scale=1.0, dtype=torch.float32):
```
Comment (Simontwice) [inline lizrd/core/init.py:5]: I don't like the default uniform here. The defaults should be moved to the config files.

File: lizrd/core/init.py (hunk 9 above)
Comment (Simontwice) [reply lizrd/core/init.py:17]: Please link to the torch PR that justifies this

File: lizrd/core/llm.py [hunk 14]
```diff
@@ -170,7 +191,16 @@ def attention_mechanism(
...
 @ash.check("... d -> ... d")
 class Attention(LoggingLayer):
-    def __init__(self, dmodel, heads, causal, dhead=None, flash=False):
+    def __init__(
+        self,
+        dmodel,
+        heads,
+        causal,
+        dhead=None,
+        flash=False,
```
Comment (Simontwice) [inline lizrd/core/llm.py:200]: delete defaults here

File: lizrd/core/llm.py [hunk 15]
```diff
@@ -170,7 +191,16 @@ def attention_mechanism(
...
+        init_type: Literal["uniform", "truncated_normal"] = "uniform",
+        init_scale: float = 1.0,
```
Comment (Simontwice) [inline lizrd/core/llm.py:202]: no defaults

**Review by Simontwice — CHANGES_REQUESTED** [review]
PR can be split into two separate PRs: refactor and LegacyAttention removal

File: lizrd/core/llm.py [hunk 16]
```diff
@@ -374,15 +357,43 @@ def get_current_device(self, block_num):
 
 
 @ash.check("... -> ... d")
-def TokenEmbedding(vocab_size, embedding_dim):
+def TokenEmbedding(
+    vocab_size,
+    embedding_dim,
+    init_type: Literal["uniform", "truncated_normal"] = "truncated_normal",
+    init_scale: float = 1.0,
```
Comment (Simontwice) [inline lizrd/core/llm.py:364]: no defaults

File: lizrd/core/llm.py [hunk 17]
```diff
@@ -374,15 +357,43 @@ def get_current_device(self, block_num):
 
 
 @ash.check("... -> ... d")
-def TokenEmbedding(vocab_size, embedding_dim):
+def TokenEmbedding(
+    vocab_size,
+    embedding_dim,
+    init_type: Literal["uniform", "truncated_normal"] = "truncated_normal",
```
Comment (Simontwice) [inline lizrd/core/llm.py:363]: no defaults

File: lizrd/core/llm.py [hunk 18]
```diff
@@ -374,15 +357,43 @@ def get_current_device(self, block_num):
...
+    init_scale: float = 1.0,
+):
+    embedding = nn.Embedding(vocab_size, embedding_dim)
+    default_weight = embedding.weight.data
+    embedding.weight.data = get_init_weight(
+        shape=default_weight.shape,
+        fan_in=1,
```
Comment (Simontwice) [inline lizrd/core/llm.py:370]: maybe leave some kind of comments that this is how its done in pytorch nn.Embedding

File: lizrd/core/llm.py [hunk 19]
```diff
@@ -315,15 +357,43 @@ def get_current_device(self, block_num):
...
 
 @ash.check("... -> ... d")
 class PositionalEmbedding(nn.Module):
-    def __init__(self, max_length, embedding_dim):
+    def __init__(
+        self,
+        max_length,
+        embedding_dim,
+        init_type: Literal["uniform", "truncated_normal"] = "truncated_normal",
+        init_scale: float = 1.0,
```
Comment (Simontwice) [inline lizrd/core/llm.py:385]: no likey

File: lizrd/core/misc.py [hunk 20]
```diff
@@ -93,15 +80,17 @@ def DenseEinMix(dinp, dout):
 
 @ash.check("... inp -> ... out")
 class Linear(nn.Linear):
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, init_type="uniform", init_scale=1.0, **kwargs):
```
Comment (Simontwice) [inline lizrd/core/misc.py:83]: no defaults

File: research/conditional/moe_layers/cont_moe_designs/separate_weighted_parameter_matrices.py (hunk 5 above)
Comment (Simontwice) [reply research/conditional/moe_layers/cont_moe_designs/separate_weighted_parameter_matrices.py:5]: agree

File: research/conditional/moe_layers/expert_choice.py [hunk 21]
```diff
@@ -27,6 +27,8 @@ def __init__(
         use_full_einsum: bool = False,
         softmax_over: Literal["tokens", "experts"] = "tokens",
         n_gating_heatmaps: int = 4,
+        init_type: Literal["uniform", "truncated_normal"] = "uniform",
```
Comment (Simontwice) [inline research/conditional/moe_layers/expert_choice.py:30]: defaults bad

File: lizrd/core/llm.py [hunk 22]
```diff
@@ -32,17 +35,34 @@ def FeedForward(
     dmodel,
     dff,
     bias: Literal["both", "first", "second", "none"] = "both",
+    init_type: Literal["uniform", "truncated_normal"] = "uniform",
```
Comment (Simontwice) [inline lizrd/core/llm.py:38]: defaults bad

**Review by Simontwice — CHANGES_REQUESTED** [review]
the defaults in init scale/type are unsafe - it is safer to have no defaults and set it in the config file

File: lizrd/core/init.py (hunk 13 above)
Comment (sebastianjaszczur) [reply lizrd/core/init.py:5]: +1

File: lizrd/core/init.py (hunk 12 above)
Comment (sebastianjaszczur) [reply lizrd/core/init.py:6]: +1

File: lizrd/core/init.py (hunk 9 above)
Comment (sebastianjaszczur) [reply lizrd/core/init.py:17]: also, do we care about relu in anyway? If we use relu after a linear projection we should, I think, multiply the scale by a factor of two, maybe? altho ignoring it completely sounds fine as well

Comment (sebastianjaszczur) [reply lizrd/core/init.py:17]: <jk, unless, but joking, but maybe we should> We can also replace relu with ABSOLUTE VALUE for maximum efficiency and to keep the magnitude constant (altho sadly it won't allow for in-place operations) <\jk, unless, but joking, but maybe we should>

File: research/conditional/tests/test_token_choice.py (hunk 3 above)
Comment (sebastianjaszczur) [reply research/conditional/tests/test_token_choice.py:7]: +1

**Review by sebastianjaszczur — COMMENTED** [review]
In general I agree with others, I didn't notice anything major apart from existing comments. Thanks for the PR! I think it'll allow for way better tuning of initialization as well :)

File: lizrd/core/init.py (hunk 9 above)
Comment (Simontwice) [reply lizrd/core/init.py:17]: I very much support this notion

Comment (kuba-krj) [reply lizrd/core/init.py:17]: After discussion we leave it as is

**Review by kuba-krj — COMMENTED** [review]
I addressed your comments. Link to [Neptune](https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9a723c30-f1d9-4f65-bc13-fc80df95da12&detailsTab=monitoring&shortId=LLMRANDOM-2308&dash=charts&type=run&compare=EwZgDALANCXEA&chartFilter=loss) with comparison of old/new initialization

File: lizrd/core/init.py [hunk 23]
```diff
@@ -0,0 +1,34 @@
+import torch
+from torch.nn.init import trunc_normal_
+
+
+def get_init_weight(shape, fan_in, init_type, scale=1.0, dtype=torch.float32):
```
Comment (Simontwice) [inline lizrd/core/init.py:5]: should scale remain a default?

File: lizrd/core/llm.py [hunk 24]
```diff
@@ -31,18 +33,35 @@ def decode_bias_string(bias):
 def FeedForward(
     dmodel,
     dff,
+    init_type: Literal["kaiming_uniform", "truncated_normal"],
     bias: Literal["both", "first", "second", "none"] = "both",
+    init_scale: float = 1.0,
```
Comment (Simontwice) [inline lizrd/core/llm.py:38]: as above

File: lizrd/core/init.py
Comment (crewtool) [inline lizrd/core/init.py:1]: maybe rename to initialization, as init is too similar to __init__

**Review by crewtool — APPROVED** [review]
LGTM

**Review by Simontwice — CHANGES_REQUESTED** [review]
apart from minor changes, LGTM

File: lizrd/core/init.py
Comment (kuba-krj) [reply lizrd/core/init.py:1]: I don't understand

File: research/conditional/moe_layers/cont_moe_designs/separate_merge_emit_weights.py [hunk 25]
```diff
@@ -10,20 +11,18 @@ class ContinuousMoEMergeDifferentlySimple(ContinuousMoeBaseClass):
 
     def init_parameters(self):
         self.lin1 = nn.Parameter(
-            misc.get_init_weight(
-                (self.dm, self.n_experts, self.expert_size), fan_in=self.dm
-            )
+            get_init_weight((self.dm, self.n_experts, self.expert_size), fan_in=self.dm)
```
[the remark this answers is not in the document]
Comment (kuba-krj) [reply research/conditional/moe_layers/cont_moe_designs/separate_merge_emit_weights.py:14]: changed

**Review by Simontwice — APPROVED** [review]
